### ISSUE TITLE
test

### ISSUE THREAD

**Issue by 0redbluefin0** [0redbluefin0/suyo#1]
# -*- coding: utf-8 -*-
"""Colab 시작하기

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/notebooks/intro.ipynb
"""

# Commented out IPython magic to ensure Python compatibility.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM
from tensorflow.keras.callbacks import EarlyStopping

# %matplotlib inline

from io import StringIO

csv_data = """year,sel,chu,length,login,wait,supply,book
2022,1,0,6,"546600","281575","530398","277096"
2022,0,1,5,"868063","830998","882524","426268"
2023,1,0,5,"409725","154224","881789","395208"
2023,0,1,7,"749771","617154","1251869","629897"
2024,1,0,5,"536922","420089","800153","475207"
2024,0,1,6,"740890","505201","985627","638671"
2025,1,0,10,"1099826","479010","1625660","983531"
"""

data = pd.read_csv(StringIO(csv_data))
data

feature_cols = ['year', 'sel', 'chu', 'length', 'wait', 'supply', 'book']
target_col = 'login'

X = data[feature_cols].values.astype(float)
y = data[target_col].values.astype(float).reshape(-1, 1)

X, y

scaler_X = MinMaxScaler()
scaler_y = MinMaxScaler()

X_scaled = scaler_X.fit_transform(X)
y_scaled = scaler_y.fit_transform(y)

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y_scaled, test_size=0.3, random_state=42
)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

mlp_model = Sequential([
    Dense(16, activation='relu', input_shape=(X_train.shape[1],)),
    Dense(8, activation='relu'),
    Dense(1)  # 스케일된 login 예측
])

mlp_model.compile(optimizer='adam', loss='mse')
mlp_model.summary()

early_stop = EarlyStopping(
    monitor='val_loss',
    patience=50,
    restore_best_weights=True
)

history_mlp = mlp_model.fit(
    X_train, y_train,
    validation_data=(X_test, y_test),
    epochs=1000,
    batch_size=2,
    callbacks=[early_stop],
    verbose=0
)

plt.plot(history_mlp.history['loss'], label='train')
plt.plot(history_mlp.history['val_loss'], label='val')
plt.title('MLP 학습 곡선')
plt.legend()
plt.grid(True)
plt.show()

y_test_pred_scaled = mlp_model.predict(X_test)
y_test_inv = scaler_y.inverse_transform(y_test)
y_test_pred_inv = scaler_y.inverse_transform(y_test_pred_scaled)

print("=== MLP 예측 결과(테스트셋) ===")
print("실제 login :", y_test_inv.flatten())
print("예측 login :", y_test_pred_inv.flatten())

y_all_pred_scaled = mlp_model.predict(X_scaled)
y_all_pred = scaler_y.inverse_transform(y_all_pred_scaled)

result_mlp = data.copy()
result_mlp['pred_login_mlp'] = y_all_pred

result_mlp

# 추석 데이터만 필터
chuseok = data[(data['chu'] == 1)]

print("=== 추석 데이터 ===")
print(chuseok[['year','sel','chu','length','login','wait','supply','book']])

# 추석에서 wait/supply, book/supply 비율 평균 계산 (대략적인 가정)
chuseok['wait_supply_ratio'] = chuseok['wait'] / chuseok['supply']
chuseok['book_supply_ratio'] = chuseok['book'] / chuseok['supply']

mean_wait_ratio = chuseok['wait_supply_ratio'].mean()
mean_book_ratio = chuseok['book_supply_ratio'].mean()

print("평균 wait/supply 비율:", mean_wait_ratio)
print("평균 book/supply 비율:", mean_book_ratio)

year_new = 2025
sel_new = 0
chu_new = 1
length_new = 11
supply_new = 2144741  # 주신 값

wait_new = supply_new * mean_wait_ratio
book_new = supply_new * mean_book_ratio

print("가정한 2025 추석 wait:", wait_new)
print("가정한 2025 추석 book:", book_new)

# 새 샘플(2025 추석) 입력 벡터
X_new = np.array([[year_new, sel_new, chu_new, length_new, wait_new, supply_new, book_new]])

# 기존 스케일러로 변환
X_new_scaled = scaler_X.transform(X_new)

# 예측 (스케일된 login)
y_new_pred_scaled = mlp_model.predict(X_new_scaled)
y_new_pred = scaler_y.inverse_transform(y_new_pred_scaled)

print("=== 2025년 추석(연휴 11일, supply=2,144,741) MLP 예측 결과 ===")
print("예상 login 수:", float(y_new_pred[0,0]))
